### PR TITLE
Sast Lightweight Scan Add New Command (AST-42347)

### PR DESCRIPTION
**Pull request by elchananarb**
### Description
JS Wrapper | Sast Lightweight Scan Add New Command

 
### References

> https://checkmarx.atlassian.net/browse/AST-42347
> https://checkmarx.atlassian.net/wiki/spaces/AID/pages/7349010436

### CODE REVIEW

File: src/main/vorpal/CxVorpal.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        scan.status = resultObject.status;
+        scan.message = resultObject.message;
+        scan.error = resultObject.error;
+
+        if (resultObject.scan_details instanceof Array) {
+            scan.scan_details = resultObject.scan_details.map((detail: any) => {
+                const scanDetail = new VorpalScanDetail();
+                scanDetail.rule_id = detail.rule_id;
+                scanDetail.language = detail.language;
+                scanDetail.query_name = detail.query_name;
```
Comment (AlvoBen) [inline src/main/vorpal/CxVorpal.ts:30]: I looked at the proto file, there is no QueryName. I think is supposed to be RuleName.
Also, there should be another field: ProblematicLine.

File: src/main/vorpal/VorpalScanDetail.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+export default class VorpalScanDetail {
+    rule_id: number;
+    language: string;
+    query_name: string;
+    severity: string;
+    file_name: string;
+    line: number;
+    length: number;
+    remediation: string;
+    description: string;
```
Comment (AlvoBen) [inline src/main/vorpal/VorpalScanDetail.ts:10]: add those fields here

File: src/main/wrapper/ExecutionService.ts [hunk 3]
```diff
@@ -183,6 +184,10 @@ export class ExecutionService {
                   const scans = CxScan.parseProject(resultObject);
                   cxCommandOutput.payload = scans;
                   break;
+                case CxConstants.SCAN_VORPAL:
+                    const vorpal = CxVorpal.parseScan(resultObject);
```
Comment (OrShamirCM) [inline src/main/wrapper/ExecutionService.ts:188]: If the CLI returns an error (not the object) Its should know how to handle it too and parse to a string then pass tit.

File: src/tests/ScanTest.test.ts [hunk 4]
```diff
@@ -125,4 +125,69 @@ describe("ScanCreate cases", () => {
...
+        {
+            rule_id: 0,
+            language: "Python",
+            query_name: "Stored XSS",
+            severity: "High",
+            file_name: "python-vul-file.py",
+            line: 37,
+            length: 0,
+            remediation: "Fully encode all dynamic data, regardless of source, before embedding it in output.",
+            description: "The method undefined embeds untrusted data in generated output with write, at line 80 of /dsvw.py. This untrusted data is embedded into the output without proper sanitization or encoding, enabling an attacker to inject malicious code into the generated web-page. The attacker would be able to alter the returned web page by saving malicious data in a data-store ahead of time. The attacker's modified data is then read from the database by the undefined method with read, at line 37 of /dsvw.py. This untrusted data then flows through the code straight to the output web page, without sanitization.  This can enable a Stored Cross-Site Scripting (XSS) attack."
```
Comment (OrShamirCM) [inline src/tests/ScanTest.test.ts:178]: Is this the mock repsonse asser or an assert that can be generic?
We should have a test for error too

**Review by OrShamirCM — APPROVED** [review]

File: src/main/wrapper/CxWrapper.ts [hunk 5]
```diff
@@ -125,6 +125,22 @@ export class CxWrapper {
         return await exec.executeCommands(this.config.pathToExecutable, commands, CxConstants.SCAN_TYPE);
     }
 
+    async scanVorpal(sourceFile: string, updateVersion = false, agent?: string | null): Promise<CxCommandOutput> {
```
Comment (tamarleviCm) [inline src/main/wrapper/CxWrapper.ts:128]: the agent must be provided by the plugin. do not allow 'null' agent

File: src/main/wrapper/CxWrapper.ts [hunk 6]
```diff
@@ -125,6 +125,22 @@ export class CxWrapper {
...
+        const commands: string[] = [CxConstants.CMD_SCAN, CxConstants.CMD_VORPAL, CxConstants.SOURCE_FILE, sourceFile];
+
+        if (updateVersion) {
+            commands.push(CxConstants.VORPAL_UPDATE_VERSION);
+        }
+        if (agent) {
```
Comment (tamarleviCm) [inline src/main/wrapper/CxWrapper.ts:134]: if the agent is null - send 'wrapper' or 'js-wrapper' in the flag

**Comment by AlvoBen** [comment]
Logo
**Checkmarx One – Scan Summary & Details** – [7c5499a8\-4e41\-480c\-b34e\-c6da04e3b900](https://eu.ast.checkmarx.net/projects/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/scans?id=7c5499a8-4e41-480c-b34e-c6da04e3b900&branch=feature%2Felchanan%2Fsast-lightweight-scan-add-new-command)
### Policy Management Violations \- Pull/merge request blocked
Policy Name|Rule(s)|Break Build
---|---|---
Phoenix\-Policy|1\-sca\-vul|true
### New Issues
Severity|Issue|Source File / Package|Checkmarx Insight
---|---|---|---
HIGH|[Code_Injection](https://eu.ast.checkmarx.net/results/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/7c5499a8-4e41-480c-b34e-c6da04e3b900/sast/description/94/13646819717326216658)|/tsc/tests/data/python\-vul\-file\.py: [25](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//tsc/tests/data/python-vul-file.py#L25)|[Attack Vector](https://eu.ast.checkmarx.net/results/7c5499a8-4e41-480c-b34e-c6da04e3b900/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/sast?result-id=CbEiA6jmndm1xWMJDrzeWVzX3YU%3D&redirect=true)
HIGH|[Code_Injection](https://eu.ast.checkmarx.net/results/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/7c5499a8-4e41-480c-b34e-c6da04e3b900/sast/description/94/13646819717326216658)|/tsc/tests/data/python\-vul\-file\.py: [25](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//tsc/tests/data/python-vul-file.py#L25)|[Attack Vector](https://eu.ast.checkmarx.net/results/7c5499a8-4e41-480c-b34e-c6da04e3b900/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/sast?result-id=tmUB01du2Of%2BBcjHb6h9R3Q%2FJZY%3D&redirect=true)
HIGH|[Code_Injection](https://eu.ast.checkmarx.net/results/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/7c5499a8-4e41-480c-b34e-c6da04e3b900/sast/description/94/13646819717326216658)|/tsc/tests/data/python\-vul\-file\.py: [56](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//tsc/tests/data/python-vul-file.py#L56)|[Attack Vector](https://eu.ast.checkmarx.net/results/7c5499a8-4e41-480c-b34e-c6da04e3b900/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/sast?result-id=Cf2fBhfjoLnzD1HJjIHhxS7X9cU%3D&redirect=true)
HIGH|[Command_Injection](https://eu.ast.checkmarx.net/results/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/7c5499a8-4e41-480c-b34e-c6da04e3b900/sast/description/77/10979096389566260942)|/tsc/tests/data/python\-vul\-file\.py: [56](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//tsc/tests/data/python-vul-file.py#L56)|[Attack Vector](https://eu.ast.checkmarx.net/results/7c5499a8-4e41-480c-b34e-c6da04e3b900/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/sast?result-id=wsCw8j6lyjgKefqyWaaMImxNlKc%3D&redirect=true)
HIGH|[Command_Injection](https://eu.ast.checkmarx.net/results/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/7c5499a8-4e41-480c-b34e-c6da04e3b900/sast/description/77/10979096389566260942)|/tsc/tests/data/python\-vul\-file\.py: [25](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//tsc/tests/data/python-vul-file.py#L25)|[Attack Vector](https://eu.ast.checkmarx.net/results/7c5499a8-4e41-480c-b34e-c6da04e3b900/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/sast?result-id=2BhpqAXGYzoUiWSA4dxUg4W%2Fe54%3D&redirect=true)
HIGH|[Command_Injection](https://eu.ast.checkmarx.net/results/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/7c5499a8-4e41-480c-b34e-c6da04e3b900/sast/description/77/10979096389566260942)|/tsc/tests/data/python\-vul\-file\.py: [25](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//tsc/tests/data/python-vul-file.py#L25)|[Attack Vector](https://eu.ast.checkmarx.net/results/7c5499a8-4e41-480c-b34e-c6da04e3b900/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/sast?result-id=YCcnDNjartB0hD3skces7b8aXZk%3D&redirect=true)
HIGH|[Stored_XSS](https://eu.ast.checkmarx.net/results/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/7c5499a8-4e41-480c-b34e-c6da04e3b900/sast/description/79/346558629760677672)|/tsc/tests/data/python\-vul\-file\.py: [37](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//tsc/tests/data/python-vul-file.py#L37)|[Attack Vector](https://eu.ast.checkmarx.net/results/7c5499a8-4e41-480c-b34e-c6da04e3b900/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/sast?result-id=FlTM7%2BV6fvh8TsVlpXmGzghfwtE%3D&redirect=true)
HIGH|[Stored_XSS](https://eu.ast.checkmarx.net/results/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/7c5499a8-4e41-480c-b34e-c6da04e3b900/sast/description/79/346558629760677672)|/tsc/tests/data/python\-vul\-file\.py: [35](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//tsc/tests/data/python-vul-file.py#L35)|[Attack Vector](https://eu.ast.checkmarx.net/results/7c5499a8-4e41-480c-b34e-c6da04e3b900/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/sast?result-id=hzGxTPs36MpMTKkJ2mDcdo6hrv4%3D&redirect=true)
MEDIUM|[Missing_HSTS_Header](https://eu.ast.checkmarx.net/results/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/7c5499a8-4e41-480c-b34e-c6da04e3b900/sast/description/346/7929843929890808532)|/tsc/tests/data/python\-vul\-file\.py: [76](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//tsc/tests/data/python-vul-file.py#L76)|[Attack Vector](https://eu.ast.checkmarx.net/results/7c5499a8-4e41-480c-b34e-c6da04e3b900/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/sast?result-id=S6tfiPJkqkHwYU0iEnvhyFakXoM%3D&redirect=true)
MEDIUM|Unpinned Actions Full Length Commit SHA|/ast\-scan\.yml: [12](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//.github/workflows/ast-scan.yml#L12)|Pinning an action to a full length commit SHA is currently the only way to use an action as an immutable release. Pinning to a particular SHA helps...
MEDIUM|Unpinned Actions Full Length Commit SHA|/release\.yml: [112](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//.github/workflows/release.yml#L112)|Pinning an action to a full length commit SHA is currently the only way to use an action as an immutable release. Pinning to a particular SHA helps...
MEDIUM|Unpinned Actions Full Length Commit SHA|/dependabot\-auto\-merge\.yml: [23](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//.github/workflows/dependabot-auto-merge.yml#L23)|Pinning an action to a full length commit SHA is currently the only way to use an action as an immutable release. Pinning to a particular SHA helps...
MEDIUM|Unpinned Actions Full Length Commit SHA|/release\.yml: [99](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//.github/workflows/release.yml#L99)|Pinning an action to a full length commit SHA is currently the only way to use an action as an immutable release. Pinning to a particular SHA helps...
MEDIUM|Unpinned Actions Full Length Commit SHA|/dependabot\-auto\-merge\.yml: [14](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//.github/workflows/dependabot-auto-merge.yml#L14)|Pinning an action to a full length commit SHA is currently the only way to use an action as an immutable release. Pinning to a particular SHA helps...
MEDIUM|Unpinned Actions Full Length Commit SHA|/release\.yml: [137](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//.github/workflows/release.yml#L137)|Pinning an action to a full length commit SHA is currently the only way to use an action as an immutable release. Pinning to a particular SHA helps...
MEDIUM|Unpinned Actions Full Length Commit SHA|/release\.yml: [119](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//.github/workflows/release.yml#L119)|Pinning an action to a full length commit SHA is currently the only way to use an action as an immutable release. Pinning to a particular SHA helps...
MEDIUM|Unpinned Actions Full Length Commit SHA|/delete\-packages\-and\-releases\.yml: [44](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//.github/workflows/delete-packages-and-releases.yml#L44)|Pinning an action to a full length commit SHA is currently the only way to use an action as an immutable release. Pinning to a particular SHA helps...
MEDIUM|Unpinned Actions Full Length Commit SHA|/update\-cli\.yml: [32](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//.github/workflows/update-cli.yml#L32)|Pinning an action to a full length commit SHA is currently the only way to use an action as an immutable release. Pinning to a particular SHA helps...
LOW|[Unsafe_Use_Of_Target_blank](https://eu.ast.checkmarx.net/results/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/7c5499a8-4e41-480c-b34e-c6da04e3b900/sast/description/1022/17155209939029667221)|/tsc/tests/data/python\-vul\-file\.py: [63](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//tsc/tests/data/python-vul-file.py#L63)|[Attack Vector](https://eu.ast.checkmarx.net/results/7c5499a8-4e41-480c-b34e-c6da04e3b900/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/sast?result-id=MGC%2BCXuTnfbJ5uVyV3Rk129yFSw%3D&redirect=true)
LOW|[Unsafe_Use_Of_Target_blank](https://eu.ast.checkmarx.net/results/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/7c5499a8-4e41-480c-b34e-c6da04e3b900/sast/description/1022/17155209939029667221)|/tsc/tests/data/python\-vul\-file\.py: [10](https://github.com/CheckmarxDev/ast-cli-javascript-wrapper/blob/feature/elchanan/sast-lightweight-scan-add-new-command//tsc/tests/data/python-vul-file.py#L10)|[Attack Vector](https://eu.ast.checkmarx.net/results/7c5499a8-4e41-480c-b34e-c6da04e3b900/800f0cfd-da0f-47f5-b9db-feb25b4bc73a/sast?result-id=l73Xw4eVFZWftX7uqs3IVx8xZCg%3D&redirect=true)

**Pull request merged**